[16.0][IMP]  - css sticky table header and sticky first column added

dora-jurcevic opened the pull request:
Thanks for your pull request. Please take a moment to review if it meets the following
guidelines.

## Description

Following the discussion on the PR for the same feature on V14 https://github.com/OCA/mis-builder/pull/588.
The PR adds a sticky header and first column on the Mis budget report.
On the first gif you can see the vertical scroll, on the second one the horizontal one. 

![mis-builder-vertical-scroll](https://github.com/user-attachments/assets/cb38f97e-eab4-425d-bf99-357e228c1372)
![mis-builder-sticky-header-horizontal-scroll](https://github.com/user-attachments/assets/573a2371-0e84-41a0-8353-72ff2ac08a52)

Thx !

vdewulf commented:
Hello,
I tested (not on runboat) on a copy of a database we have with a wide report already configured. Unfortunately, the display is not so good, as you can see:
![Sélection_030](https://github.com/user-attachments/assets/cc96ab46-934e-4011-a5fc-f10f9bd2af66)

For comparison, this is the report with non sticky header in production database:
![Sélection_031](https://github.com/user-attachments/assets/42be0fbb-df5e-4ca9-afe3-055708e58040)

The stickyness is working for the title line, but the whole rendering is not working. 
Is it something in my database?!

sbidoul commented:
From a technical standpoint, I worry that these overrides of standard odoo css classes would have side effects in other contexts than mis builder? Not sure.

Also, is there a better solution than hard coding colors?

dora-jurcevic commented:
Thank you all for the feedback. 
Ia I was advised, I decided to split in 2 different PRs the two features : sticky header + sticky first column. 
I added a commit with a refactor keeping only the sticky header in this PR. I hope it would be easier to test.

@sbidoul , thank you, I changed the colors by using the standard odoo bootstrap variables.
Also, to reinsure that the classes don't interfere or override classes on other tables / views, I further nested the css classes with those only used on miss builder , for example : 
`.oe_mis_builder_content .o_list_renderer thead th {
    position: -webkit-sticky;
    position: sticky;
    top: 0;
    background-color: var(--gray-300);
    z-index: 1;
    border: 0.2px solid var(--gray-100);
  }`

Here the style applies only on html table header elements where the parent has _oe_mis_builder_content_ class (and o_list_renderer ) applied . This ensures that the style won't apply on other odoo tables or lists. 
@vdewulf , would be great to see if my refactor works on your reports !
Thank you for you collaboration and testing !

sbidoul commented:
Thank you Dora. Can you update the PR title to reflect this?

sbidoul commented on mis_builder/static/src/components/mis_report_widget.css:
```diff
@@ -1,3 +1,22 @@
+.oe_mis_builder_content {
+    max-width: 98vw;
+}
```
Is this change necessary for the sticky header? Or is it a remnant of the horizontal scroll attempt? If the latter, let's remove it.

sbidoul commented on mis_builder/static/src/components/mis_report_widget.css:
```diff
@@ -1,3 +1,22 @@
+    position: sticky;
+    top: 0;
+    background-color: var(--gray-300);
+    z-index: 1;
+    border: 0.2px solid var(--gray-100);
```
In my tests (firefox & chrome on linux), z-index and border don't make much difference. Are they important in other browsers?

sbidoul commented on mis_builder/static/src/components/mis_report_widget.css:
```diff
@@ -1,3 +1,22 @@
+}
+
+.oe_mis_builder_content .o_list_renderer thead th {
+    position: -webkit-sticky; /* Pour Safari */
```
Is `-webkit-sticky` still necessary nowadays?

dora-jurcevic replied:
I did a research and it seems that is still is not fully supported  : https://www.lambdatest.com/web-technologies/css-sticky-safari. Also MDN uses it in an example : https://developer.mozilla.org/en-US/docs/Web/CSS/position. I would be keen on keeping it

sbidoul reviewed (commented):
A few minor questions. 

Can you also fix the pre-commit issues?

sbidoul commented on mis_builder/static/src/components/mis_report_widget.css:
```diff
@@ -1,3 +1,22 @@
+    width: 100%;
+    overflow: auto;
+    position: relative;
+    height: 95vh; /*limit the height in order for the vertical scroll to work  */
```
In my tests I need to set it to 80vh to have the full report appear and have the horizontal scroll bar show up at the bottom.

dora-jurcevic replied:
hello @sbidoul thank you for testing ! 
Not sure to see exactly what you mean ? 
The height limit is to assure the vertical scroll appears if the table is smaller than the screen height.
Could  you please add a screenshot with the 80vh modification ?

sbidoul replied:
@dora-jurcevic I mean this:

![image](https://github.com/user-attachments/assets/ff18a44a-43ae-4821-b6b8-c330132a3e47)

vdewulf commented:
@dora-jurcevic I tested the last version on the runboat here and it doesn't seem to work.
See my video.
![mis_report_sticky_header_test_runboat_18042025](https://github.com/user-attachments/assets/dacb9bd6-41d7-431c-8872-e9b0a4f930f5)

dora-jurcevic commented:
> @dora-jurcevic I tested the last version on the runboat here and it doesn't seem to work. See my video. ![mis_report_sticky_header_test_runboat_18042025](https://private-user-images.githubusercontent.com/27895270/435240621-dacb9bd6-41d7-431c-8872-e9b0a4f930f5.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ5OTE0MDIsIm5iZiI6MTc0NDk5MTEwMiwicGF0aCI6Ii8yNzg5NTI3MC80MzUyNDA2MjEtZGFjYjliZDYtNDFkNy00MzFjLTg4NzItZTliMGE0ZjkzMGY1LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDE4VDE1NDUwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMwZTk2MDAxMGZkYjQ4YTQyNzEwYTUzZTdjMGI5MDVhZWUwNmQ4Y2JiMzdkNTVhMzIxODcyN2Q0M2UzZTllZjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.AVzDxvL7FIyV64wPw7ex0ci4pt3-25pmnbIU2e8UVW4) [ ![mis_report_sticky_header_test_runboat_18042025](https://private-user-images.githubusercontent.com/27895270/435240621-dacb9bd6-41d7-431c-8872-e9b0a4f930f5.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ5OTE0MDIsIm5iZiI6MTc0NDk5MTEwMiwicGF0aCI6Ii8yNzg5NTI3MC80MzUyNDA2MjEtZGFjYjliZDYtNDFkNy00MzFjLTg4NzItZTliMGE0ZjkzMGY1LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDE4VDE1NDUwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMwZTk2MDAxMGZkYjQ4YTQyNzEwYTUzZTdjMGI5MDVhZWUwNmQ4Y2JiMzdkNTVhMzIxODcyN2Q0M2UzZTllZjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.AVzDxvL7FIyV64wPw7ex0ci4pt3-25pmnbIU2e8UVW4) ](https://private-user-images.githubusercontent.com/27895270/435240621-dacb9bd6-41d7-431c-8872-e9b0a4f930f5.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ5OTE0MDIsIm5iZiI6MTc0NDk5MTEwMiwicGF0aCI6Ii8yNzg5NTI3MC80MzUyNDA2MjEtZGFjYjliZDYtNDFkNy00MzFjLTg4NzItZTliMGE0ZjkzMGY1LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDE4VDE1NDUwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMwZTk2MDAxMGZkYjQ4YTQyNzEwYTUzZTdjMGI5MDVhZWUwNmQ4Y2JiMzdkNTVhMzIxODcyN2Q0M2UzZTllZjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.AVzDxvL7FIyV64wPw7ex0ci4pt3-25pmnbIU2e8UVW4) [ ](https://private-user-images.githubusercontent.com/27895270/435240621-dacb9bd6-41d7-431c-8872-e9b0a4f930f5.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ5OTE0MDIsIm5iZiI6MTc0NDk5MTEwMiwicGF0aCI6Ii8yNzg5NTI3MC80MzUyNDA2MjEtZGFjYjliZDYtNDFkNy00MzFjLTg4NzItZTliMGE0ZjkzMGY1LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDE4VDE1NDUwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMwZTk2MDAxMGZkYjQ4YTQyNzEwYTUzZTdjMGI5MDVhZWUwNmQ4Y2JiMzdkNTVhMzIxODcyN2Q0M2UzZTllZjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.AVzDxvL7FIyV64wPw7ex0ci4pt3-25pmnbIU2e8UVW4)

Thank you @vdewulf  for testing. So on the OCA bot the testing is not really ideal beacuse the table has only 3 rows. For the sticky header to appear, the table should be bigger (heigher) than the screen . You can "cheat" by : 
- inspecting the table
- selecting a <tr> element in the inspector tool
- right click -> duplicate node 
- this will copy a row , and if you do it 15-20 times (easy peasy ;) ) , you can simulate a big table without having to reproduce enough data for a big report.

Hope this helps !

hitrosol commented:
I added many kpi's row in the runboat http://oca-mis-builder-16-0-pr673-458d6234f03c.runboat.odoo-community.org/web#id=1&cids=1&menu_id=120&action=281&model=mis.report&view_type=form
the sticky header is working.

But i found that the report is squeezed in one page, this give difficult views if there are many columns in one report.
It be good if the report can be freezed in the first column as well. 
thanks

vdewulf commented:
@dora-jurcevic Test successfull on my customer database.

As stated together previously, as a next improvement, the first column should be fixed, but with an attention to fix the width of the column because some of our customers have really (really) long names of accounts, as you will see in this GIF.
![mis_report_sticky_header_test_real-database_01052025](https://github.com/user-attachments/assets/4210272f-9b44-4e78-a63f-cbe81a55c822)

You need to fix the pre-commit aspects (no idea how) before we can merge this, after answering the last comments from @sbidoul. Thanks

vdewulf reviewed (approved):
Functional test ok, see my dedicated comment added today

sbidoul commented:
@dora-jurcevic I made https://github.com/akretion/mis-builder/pull/1 with the pre-commit fix and the changes I suggested above.

sbidoul commented:
> I made [akretion#1](https://github.com/akretion/mis-builder/pull/1) with the pre-commit fix and the changes I suggested above.

@dora-jurcevic can you merge [akretion#1](https://github.com/akretion/mis-builder/pull/1) if ok for you?

AnizR commented on mis_builder/static/src/components/mis_report_widget.css:
```diff
@@ -1,3 +1,19 @@
+    overflow: auto;
+    position: relative;
+    /* limit height in order for the vertical scroll to work */
+    height: 80vh;
```
I think that it would be better to use:
```suggestion
    max-height: 80vh;
```

It removes this area when having not a lot of lines and allows to vertically scroll when having multiple lines
![image](https://github.com/user-attachments/assets/0f432d66-b697-4b96-8a5a-c2a546719cca)

dora-jurcevic replied:
yes ! Thx ! I just commited the suggested change !

AnizR reviewed (approved):
Great, Code LGTM :+1:

vdewulf commented:
@sbidoul hello there! would you agree to merge this PR?

sbidoul commented:
I agree, but a git rebase is needed.

sbidoul commented:
I can do it in anoter PR if you prefer, preserving your authorship (of course).

dora-jurcevic commented:
hello @sbidoul ! Please see this MR, I rebased the code and everything seems to be green now. 
Thx !

sbidoul commented:
Thank you Dora!

/ocabot merge minor
